tools/internal/parser/text.go

```diff
@@ -0,0 +1,141 @@
+package parser
+
+import (
+	"strings"
+)
+
+// source is a block of zero or more lines of source text. It can be
+// decomposed into smaller sub-blocks using its parsing methods. It
+// also preserves source line information and can be turned into an
+// exported Source when needed.
+type source struct {
+	// The lines of source text.
+	lines []string
+	// lineOffset is how many lines are before the beginning of lines,
+	// for sources that represent a subset of the input.
+	lineOffset int
+}
+
+// newSource returns a source for bs.
+func newSource(src string) source {
+	return source{
+		lines:      strings.Split(src, "\n"),
+		lineOffset: 0,
+	}
+}
+
+// Source returns the exported Source counterpart of s.
+func (s source) Source() Source {
+	if len(s.lines) == 0 {
+		// Letting a zero-byte source range escape from the parser
+		// package is the sign of a bug somewhere, so blow up and make
+		// it visible.
+		panic("can't construct a zero-line Source")
+	}
+	start := s.lineOffset + 1
+	end := start + len(s.lines) - 1
+	return Source{
+		StartLine: start,
+		EndLine:   end,
+		Raw:       s.Text(),
+	}
+}
+
+// slice returns the slice of s between startLine and endLine.
+//
+// startLine and endLine behave like normal slice offsets, i.e. they
+// represent the half-open range [startLine:endLine).
```

Comment:
So using StartLine and EndLine from a Source won't work here because there the EndLine is included, right?